### ISSUE TITLE
Weird behavior of custom inference.py script

### ISSUE THREAD

**Issue by la-cruche** [aws/sagemaker-huggingface-inference-toolkit#5]
Hi, 

I'm deploying the custom inference.py script below

```python
import os

import tensorflow as tf
from transformers import TFAutoModelForQuestionAnswering, AutoTokenizer




def load_fn(model_dir):
    """this function reads the model from disk"""
    
    print('load_fn dir view:')
    print(os.listdir())
    
    # load model
    model = TFAutoModelForQuestionAnswering.from_pretrained(os.environ['MODEL_FOLDER'])
    
    # load tokenizer
    tokenizer = AutoTokenizer.from_pretrained(os.environ['TOKENIZER_FOLDER'])
    
    return model, tokenizer




def preprocess_fn(input_data, content_type):
    """this function pre-processes the input.
       payload format is {"inputs": {"question": XX, "text": XX}}"""
    
    print('input_data received:')
    print(input_data)
    return input_data['inputs']['question'], input_data['inputs']['text']




def predict(processed_data)
    """this function runs inference"""

    print('processed_data received: ')
    print(processed_data)
    
    question, text = processed_data

    input_dict = tokenizer(question, text, return_tensors='tf')
    outputs = model(input_dict)
    start_logits = outputs.start_logits
    end_logits = outputs.end_logits
    
    all_tokens = tokenizer.convert_ids_to_tokens(input_dict["input_ids"].numpy()[0])
    answer = ' '.join(all_tokens[tf.math.argmax(start_logits, 1)[0] : tf.math.argmax(end_logits, 1)[0]+1])
    
    return answer

```

There are 2 weirds things happening:

1. the print statements are not showing up in CloudWatch
2. the inference errors with `W-model-1-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - mms.service.PredictionException: invalid syntax (inference.py, line 39) : 400`

Anything wrong in the code?

**Comment by philschmid** [comment]
> 1. the print statements are not showing up in CloudWatch

@vdantu can you share your `infernece.py` where you can see the `print` statements? 

> 2. the inference errors with W-model-1-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - mms.service.PredictionException: invalid syntax (inference.py, line 39) : 400

@la-cruche what is line 39? 
Also, I am not sure if your `preprocess` function is correct since you are not parsing the JSON.  

---
Also, why are you writing your own inference code for `question-answering` and not using `HF_TASK` and let the inference-toolkit handle everything?

**Comment by la-cruche** [comment]
Hi thanks @philschmid I am parsing the input here no? 
`input_data['inputs']['question'], input_data['inputs']['text']`
my thought was that input_data would be a dict in memory. Otherwise what's the point of SageMaker knowing "content_type" if we have to do the deser ourselves for each content type?

the reason I'm doing my own inference code is because that's the thing customers are going to do the most ; they will always want to finely control what happens, what the result is (probas, strings, top-N), and they will also often chain multiple things in the endpoint (eg embedding + kNN, ensembles...). As part of bug bash I just wanted to check that this advanced yet frequent pattern would work fine :)

**Comment by philschmid** [comment]
> Hi thanks @philschmid I am parsing the input here no?
> input_data['inputs']['question'], input_data['inputs']['text']
> my thought was that input_data would be a dict in memory. Otherwise what's the point of SageMaker knowing "content_type" if we have to do the deser ourselves for each content type?

`input_fn` or `preprocess` receive the raw HTTP-Body + the `content-type` to be able to parse the body. E.g. If a customer wants to sent a `png` he could implement in the `input_fn` based on the `content-type` his preprocess function for handling the `png`.

You always need to parse the body. Take a look [here](https://github.com/aws/sagemaker-huggingface-inference-toolkit/blob/a3e85a1670ca10fb4055932afa5dcba8fe02589e/src/sagemaker_huggingface_inference_toolkit/handler_service.py#L127). Furthermore it is documented [here](https://github.com/aws/sagemaker-huggingface-inference-toolkit#-user-defined-codemodules) or [here](https://sagemaker.readthedocs.io/en/stable/frameworks/pytorch/using_pytorch.html?highlight=input_fn#process-model-input)

**Comment by la-cruche** [comment]
ok so my `predict(processed_data)` missed a `:` (!). Let me fix that and see what happens.

**Comment by la-cruche** [comment]
mm ok so with that new script below I now get a `W-model-1-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - mms.service.PredictionException: ("You need to define one of the following ['feature-extraction', 'text-classification', 'token-classification', 'question-answering', 'table-question-answering', 'fill-mask', 'summarization', 'translation', 'text2text-generation', 'text-generation', 'zero-shot-classification', 'conversational', 'image-classification'] as env 'TASK'.", 403) : 400`



```python
import os

import tensorflow as tf
from transformers import TFAutoModelForQuestionAnswering, AutoTokenizer




def load(model_dir):
    """this function reads the model from disk"""
    
    print('load_fn dir view:')
    print(os.listdir())
    
    # load model
    model = TFAutoModelForQuestionAnswering.from_pretrained(os.environ['MODEL_FOLDER'])
    
    # load tokenizer
    tokenizer = AutoTokenizer.from_pretrained(os.environ['TOKENIZER_FOLDER'])
    
    return model, tokenizer




def predict(processed_data):
    """this function runs inference"""

    print('processed_data received: ')
    print(processed_data)
    
    question, text = processed_data['input']['question'], processed_data['input']['text']

    input_dict = tokenizer(question, text, return_tensors='tf')
    outputs = model(input_dict)
    start_logits = outputs.start_logits
    end_logits = outputs.end_logits
    
    all_tokens = tokenizer.convert_ids_to_tokens(input_dict["input_ids"].numpy()[0])
    answer = ' '.join(all_tokens[tf.math.argmax(start_logits, 1)[0] : tf.math.argmax(end_logits, 1)[0]+1])
    
    return answer

```

Maybe that's because I didn't specify the task in my SDK call? Since my `config.json` (below) contains couple hints that it's doing QA I was thinking I wouldn't need to specify the task

```
{
  "_name_or_path": "distilbert-base-uncased-distilled-squad",
  "activation": "gelu",
  "architectures": [
    "DistilBertForQuestionAnswering"
  ],
  "attention_dropout": 0.1,
  "dim": 768,
  "dropout": 0.1,
  "hidden_dim": 3072,
  "initializer_range": 0.02,
  "max_position_embeddings": 512,
  "model_type": "distilbert",
  "n_heads": 12,
  "n_layers": 6,
  "pad_token_id": 0,
  "qa_dropout": 0.1,
  "seq_classif_dropout": 0.2,
  "sinusoidal_pos_embds": false,
  "tie_weights_": true,
  "transformers_version": "4.8.1",
  "vocab_size": 30522
}

```

**Comment by philschmid** [comment]
The inference toolkit will automatically detect the `TASK` based on `architectures` in the `config.json`. If you use the model without custom modules it should work. 

You receive the error because you didn't overwrite the `load` function. To overwrite the `load` function you need to define the `load_fn` not `load`. That's, why it is using the regular `load` function from the default handler. From your other issue #9 I expect that your `model_dir` structure is wrong.

**Comment by la-cruche** [comment]
ok I confirm the `print` were ignored because I was using a `load` function. With a `load_fn`, the `print` work correctly. nice!

**Comment by la-cruche** [comment]
note - the new contract for model loading is `model_fn`. above-described issue fixed here: https://github.com/aws/sagemaker-huggingface-inference-toolkit/issues/10#issuecomment-870404588
